dbus-wrapper/swupdd-main.c

```diff
@@ -0,0 +1,985 @@
+ *   it under the terms of the GNU General Public License as published by
+ *   the Free Software Foundation, version 2 or later of the License.
+ *
+ *   This program is distributed in the hope that it will be useful,
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *   GNU General Public License for more details.
+ *
+ *   You should have received a copy of the GNU General Public License
+ *   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Contact: Dmitry Rozhkov <dmitry.rozhkov@intel.com>
+ *
+ */
+
+#define _GNU_SOURCE
+
+#include <stdlib.h>
+#include <unistd.h>
+#include <stdio.h>
+#include <stdbool.h>
+#include <errno.h>
+#include <limits.h>
+#include <assert.h>
+#include <fcntl.h>
+#include <termios.h>
+#include <sys/wait.h>
+#include <systemd/sd-bus.h>
+#include <systemd/sd-daemon.h>
+
+#include "log.h"
+#include "list.h"
+
+#define SWUPD_CLIENT    "swupd"
+#define TIMEOUT_EXIT_SEC 30
+
+typedef enum {
+	METHOD_NOTSET = 0,
+	METHOD_CHECK_UPDATE,
+	METHOD_UPDATE,
+	METHOD_VERIFY,
+	METHOD_BUNDLE_ADD,
+	METHOD_BUNDLE_REMOVE,
+	METHOD_HASH_DUMP,
+	METHOD_SEARCH
+} method_t;
+
+typedef struct _daemon_state {
+	sd_bus *bus;
+	pid_t child;
+	method_t method;
+} daemon_state_t;
+
+static const char * const _method_str_map[] = {
+	NULL,
+	"CheckUpdate",
+	"Update",
+	"Verify",
+	"BundleAdd",
+	"BundleRemove",
+	"HashDump",
+	"Search"
+};
+
+static const char * const _method_opt_map[] = {
+	NULL,
+	"check-update",
+	"update",
+	"verify",
+	"bundle-add",
+	"bundle-remove",
+	"hashdump",
+	"search"
+};
+
+static char **list_to_strv(struct list *strlist)
+{
+	char **strv;
+	char **temp;
+
+	strv = (char **)calloc((list_len(strlist) + 1), sizeof(char *));
+
+	temp = strv;
+	while (strlist)
+	{
+		*temp = strlist->data;
+		temp++;
+		strlist = strlist->next;
+	}
+	return strv;
+}
+
+static int is_in_array(const char *key, char const * const arr[])
+{
+	if (arr == NULL) {
+		return 0;
+	}
+
+	char const * const *temp = arr;
+	while (*temp) {
+		if (strcmp(key, *temp) == 0) {
+			return 1;
+		}
+		temp++;
+	}
+	return 0;
+}
+
+static int bus_message_read_option_string(sd_bus_message *m,
+					  const char *optname,
+					  struct list **strlist,
+					  sd_bus_error *error)
+{
+	int r = 0;
+	char *option = NULL;
+	const char *value;
+
+	r = sd_bus_message_enter_container(m, SD_BUS_TYPE_VARIANT, "s");
+	if (r < 0) {
+		sd_bus_error_set_errnof(error, -r, "Failed to enter variant container of '%s'", optname);
+		return r;
+	}
+	r = sd_bus_message_read(m, "s", &value);
+	if (r < 0) {
+		sd_bus_error_set_errnof(error, -r, "Can't read value of '%s'", optname);
+		return r;
+	}
+	r = sd_bus_message_exit_container(m);
+	if (r < 0) {
+		sd_bus_error_set_errnof(error, -r, "Can't exit variant container of '%s'", optname);
+		return r;
+	}
+
+	r = asprintf(&option, "--%s", optname);
```

Comment:
Do you really want to be handling the ENOMEM case? It isn't something we tend to do elsewhere (mostly just abort in string_or_die).